src/misc/rwqueue.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2021-2021  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "rwqueue.h"
+
+#include <cassert>
+
+template <typename T>
+RWQueue<T>::RWQueue(size_t queue_capacity) : capacity(queue_capacity)
+{
+	assert(capacity > 0);
+}
+
+template <typename T>
+size_t RWQueue<T>::Size()
+{
+	std::lock_guard<std::mutex> lock(mutex);
+	return queue.size();
+}
+
+template <typename T>
+size_t RWQueue<T>::MaxCapacity() const
+{
+	return capacity;
+}
+
+template <typename T>
+bool RWQueue<T>::IsEmpty()
+{
+	std::lock_guard<std::mutex> lock(mutex);
+	return !queue.size();
+}
+
+template <typename T>
+void RWQueue<T>::CopyIn(const T &item)
+{
+	// wait until the queue has room to accept the item
+	std::unique_lock<std::mutex> lock(mutex);
+	while (queue.size() >= capacity)
+		has_room.wait(lock);
+
+	// add it, and notify the next waiting thread that we've got an item
+	queue.emplace(queue.end(), item);
+	lock.unlock();
+	has_items.notify_one();
+}
+
+template <typename T>
+void RWQueue<T>::MoveIn(T &item)
+{
+	// wait until the queue has room to accept the item
+	std::unique_lock<std::mutex> lock(mutex);
+	while (queue.size() >= capacity)
+		has_room.wait(lock);
+
+	// add it, and notify the next waiting thread that we've got an item
+	queue.emplace(queue.end(), std::move(item));
+	lock.unlock();
+	has_items.notify_one();
+}
+
+template <typename T>
+T RWQueue<T>::PopOut()
+{
+	// wait until the queue has an item that we can get
+	std::unique_lock<std::mutex> lock(mutex);
+	while (!queue.size())
+		has_items.wait(lock);
+
+	// get it, and notify the first waiting thread that the queue has room
+	T item = std::move(queue.front());
+	queue.pop_front();
+	lock.unlock();
+	has_room.notify_one();
+	return item;
+}
+
+// Explicit template instantiations
+template class RWQueue<int>; // Unit tests
```

Comment:
So every user is supposed to add their own instantiation in here?

Public template classes are usually implemented in headers-only (however inconvenient that is) to avoid this problem.

Reply:
> So every user is supposed to add their own instantiation in here?

Yes, specifically because we will be re-using the same type in multiple CPPs. 

For example, we already have `T` being `std::vector<int16_t>` in MT-32 and FluidSynth, so these two use cases reduce compiled code bloat by 2x (versus all-header implementation).

I can see us using the same `T is std::vector<int16_t>` in CD-DA and GUS as well, meaning code-bloat reduced by another factor of two (versus all-header implementation).

Reply:
> (…) so these two use cases reduce compiled code bloat by 2x (versus all-header implementation).

2x is overselling it… Binary size would grow by <~1.8% before stripping but now it grows by <~0.9%. Except now we pointlessly compile specialization for `int` that would not exist in release builds otherwise so we might be actually adding bloat instead of removing it.

I am somewhat sure that any modern linker for the last 10 years or so will avoid placing duplicated code that started as a template in the resulting binary. 

It might speed up the compilation speed by an insignificant amount (it would be different if we talked about a template class that's used in thousands of places… but this is used in 2 places). It might speed up or slow down linker (I have no idea).

Overall I don't care about such a tiny difference… I think eventually we'll just de-template this code so I am ok with splitting it into a header and cpp file.

> but I can also see us using the queue's as true asynchronous FIFOs

That code has a bunch of edge-cases and does not need a multithreaded queue. Just saying…

Anyway, the last thing I don't like here is that this file invents its own style for placing includes at the end of file (I hate whenever it happens in older legacy code as well). But it's not such a big deal as to prevent merging :)

Reply:
Thanks @dreamer.

If the linker bakes out redundant template code, then that would help a lot toward easing this concern on my part. (I'm going by my historical account back when linkers weren't as smart, and duplicate template instantiations had a very measurable performance impact as they hurt the processor's instruction cache.. that was a long time ago though; 😅)